File: docker/nextjs/src/lib/monitoring/metrics.ts
/**
 * CloudWatch Embedded Metric Format (EMF) ヘルパー
 *
 * enableMonitoring=true 時は EMF 形式で stdout に出力。
 * enableMonitoring=false 時は no-op（パフォーマンス影響なし）。
 *
 * 名前空間: PermissionAwareRAG/AdvancedFeatures
 */

const NAMESPACE = 'PermissionAwareRAG/AdvancedFeatures';

export interface MetricsLogger {
  putMetric(name: string, value: number, unit?: string): void;
  setDimension(key: string, value: string): void;
  flush(): void;
}

interface MetricEntry {
  name: string;
  value: number;
  unit: string;
}

class EmfMetricsLogger implements MetricsLogger {
  private dimensions: Record<string, string> = {};
  private metrics: MetricEntry[] = [];

  putMetric(name: string, value: number, unit = 'None'): void {
    this.metrics.push({ name, value, unit });
  }

  setDimension(key: string, value: string): void {
    this.dimensions[key] = value || 'unknown';
  }

  flush(): void {
    if (this.metrics.length === 0) return;
    try {
      // 空文字列ディメンション値を 'unknown' にフォールバック
      const safeDims: Record<string, string> = {};
      for (const [k, v] of Object.entries(this.dimensions)) {
        safeDims[k] = v || 'unknown';
      }

      const emf: Record<string, unknown> = {
        _aws: {
          Timestamp: Date.now(),
          CloudWatchMetrics: [{
            Namespace: NAMESPACE,
            Dimensions: Object.keys(safeDims).length > 0 ? [Object.keys(safeDims)] : [],
            Metrics: this.metrics.map(m => ({ Name: m.name, Unit: m.unit })),
          }],
        },
        ...safeDims,
      };

      for (const m of this.metrics) {
        emf[m.name] = m.value;
      }

      process.stdout.write(JSON.stringify(emf) + '\n');
    } catch {
      // stdout書き込み失敗は握りつぶし、アプリケーション処理を継続
    }
    this.metrics = [];
    this.dimensions = {};
  }
}

class NoopMetricsLogger implements MetricsLogger {
  putMetric(): void { /* no-op */ }
  setDimension(): void { /* no-op */ }
  flush(): void { /* no-op */ }
}

/**
 * MetricsLoggerを作成する。
 * @param enabled true: EMF出力、false: no-op
 */
export function createMetricsLogger(enabled: boolean): MetricsLogger {
  return enabled ? new EmfMetricsLogger() : new NoopMetricsLogger();
}

File: lambda/kb-auto-sync/metrics.py
"""
CloudWatch EMF metrics module.

CloudWatch Embedded Metric Format (EMF) を使用してカスタムメトリクスを発行する。
"""

import json
import sys
import time


def emit_metrics(
    *,
    scanned_file_count: int,
    changed_file_count: int,
    ingestion_triggered: int,
    scan_duration_ms: int,
    function_name: str,
) -> dict:
    """
    CloudWatch EMF フォーマットでメトリクスを発行する。

    Args:
        scanned_file_count: スキャンしたファイル数
        changed_file_count: 変更検出されたファイル数
        ingestion_triggered: インジェスションジョブがトリガーされたか (0 or 1)
        scan_duration_ms: スキャン所要時間（ミリ秒）
        function_name: Lambda 関数名

    Returns:
        EMF payload dict (for testing)
    """
    emf_payload = {
        "_aws": {
            "Timestamp": int(time.time() * 1000),
            "CloudWatchMetrics": [
                {
                    "Namespace": "KbAutoSync",
                    "Dimensions": [["FunctionName"]],
                    "Metrics": [
                        {"Name": "ScannedFileCount", "Unit": "Count"},
                        {"Name": "ChangedFileCount", "Unit": "Count"},
                        {"Name": "IngestionJobTriggered", "Unit": "Count"},
                        {"Name": "ScanDurationMs", "Unit": "Milliseconds"},
                    ],
                }
            ],
        },
        "FunctionName": function_name,
        "ScannedFileCount": scanned_file_count,
        "ChangedFileCount": changed_file_count,
        "IngestionJobTriggered": ingestion_triggered,
        "ScanDurationMs": scan_duration_ms,
    }

    # EMF は stdout に JSON として出力する
    print(json.dumps(emf_payload), file=sys.stdout, flush=True)

    return emf_payload
